### ISSUE TITLE
Improve data validation in finance plots

### ISSUE THREAD

**Issue by AlgoExecutor** [ScottPlot/ScottPlot#798]
Yes I know it's silly, but an empty array in `AddCandlesticks` crashes the render method :-)

Exception Message:

Exception thrown: 'System.IndexOutOfRangeException' in ScottPlot.dll
System.IndexOutOfRangeException: **Index was outside the bounds of the array.**
   at ScottPlot.Plottable.FinancePlot.GetAxisLimits()
   at ScottPlot.Settings.<>c.<AxisAutoX>b__53_4(IPlottable x)
   at System.Linq.Enumerable.WhereSelectEnumerableIterator`2.MoveNext()
   at System.Linq.Buffer`1..ctor(IEnumerable`1 source)
   at System.Linq.Enumerable.ToArray[TSource](IEnumerable`1 source)
   at ScottPlot.Settings.AxisAutoX(Int32 xAxisIndex, Double margin)
   at ScottPlot.Settings.AxisAutoUnsetAxes()
   at ScottPlot.Plot.Render(Bitmap bmp, Boolean lowQuality)
   at ScottPlot.Control.ControlBackEnd.Render(Boolean lowQuality, Boolean skipIfCurrentlyRendering)
   at ScottPlot.FormsPlot.Render(Boolean lowQuality, Boolean skipIfCurrentlyRendering)
   at DBVisor.DbVisorForm.RenderBinance(String database, String tableName) in D:\Code\DBVisor\DbVisorForm.cs:line 166

```cs
//pData is empty List
formsPlot.Plot.AddCandlesticks(pData.ToArray());
formsPlot.Plot.XAxis.DateTimeFormat(true);
formsPlot.Render();
```

Version: **ScottPlot 4.1.8-beta**

**Comment by bclehmann** [comment]
As a note, this issue also affects `AddOHLCs`.

**Comment by swharden** [comment]
We could check for `null` inside the `AddSomething()` methods, but since arrays like `xs`, `ys`, and `ohlcs` are exposed there's nothing preventing the user from instantiating them with an array but assigning null later.

Making data arrays inside plottables public fields was an early design choice, and it exchanges convenience/simplicity for safety. I've spent a lot of time re-thinking this decision, and I agree with your statement:

> It seems like we should consider on the level of the whole library

A direction I'm moving toward is not exposing the data arrays, but instead exposing public methods like `Update(OHLC[] newOhlcs)` which handle the replacements and do null checks too. In the case of scatter plots, it also allows constraining ideas like Xs and Ys should have the same number of points.

Running with this idea, perhaps the next step is to add `Update()` to `FinancePlot`. At the same time, I think I'll move toward a `List<OHLC>` under the hood so I can give finance plots an `Add(OHLC ohlc)`.

Dang, while I'm in there it seems like finance plots should have `GetSma(int n)` and `GetBollingerBands(int n)` to prevent the user from having to mess with all those functions directly.

I'll create a PR to explore these ideas. Thanks for the tip @AlgoExecutor and your input @bclehmann!

**Comment by swharden** [comment]
This refactor was a bit larger than expected, but I think I ended up with a finance plot and OHLC system which is less confusing.

* `OHLC` objects store position as `DateTime` and width as a `TimeSpan`. Since representing time using `double` is confusing for users, I hope this switch will make rendering issues related to candle width easier to fix (#812, #793)

* Finance plots no longer have a `public OHLC[] OHLCs` but instead a `private List<OHLC> OHLCs`. Validation and null checking happen when users instantiate a finance plot with data or add it later with new public `Add()` and `AddRange()` methods.

* Finance plots have `GetSMA()` and `GetBollingerBands()` methods which do all the confusing DateTime conversions under the hood and return double arrays that are ready to plot.

### Old Method (still works)

```cs
var plt = new ScottPlot.Plot(600, 400);
OHLC[] ohlcs = DataGen.RandomStockPrices(null, 75);
plt.AddCandlesticks(ohlcs);

double[] xs = DataGen.Consecutive(ohlcs.Length);

double[] sma8xs = xs.Skip(8).ToArray();
double[] sma8ys = Statistics.Finance.SMA(ohlcs, 8);
plt.AddScatter(sma8xs, sma8ys, markerSize: 0, color: Color.Blue, lineWidth: 2);

double[] sma20xs = xs.Skip(20).ToArray();
double[] sma20ys = Statistics.Finance.SMA(ohlcs, 20);
plt.AddScatter(sma20xs, sma20ys, markerSize: 0, color: Color.Navy, lineWidth: 2);
```

### New Method

```cs
var plt = new ScottPlot.Plot(600, 400);
OHLC[] ohlcs = DataGen.RandomStockPrices(null, 75);
var candlePlot = plt.AddCandlesticks(ohlcs);

var sma8 = candlePlot.GetSMA(8);
plt.AddScatterLines(sma8.xs, sma8.ys, Color.Blue, 2);

var sma20 = candlePlot.GetSMA(20);
plt.AddScatterLines(sma20.xs, sma20.ys, Color.Navy, 2);
```

### Output

![image](https://user-images.githubusercontent.com/4165489/108610321-cf046500-73a2-11eb-9966-3a6ea76c4b27.png)

**Comment by AlgoExecutor** [comment]
Unfortunately, it looks like the exception is still in there in Version: **ScottPlot 4.1.9-beta** 
An empty array in `AddCandlesticks` still crash the render methode :boom:

System.ArgumentOutOfRangeException: **Index was out of range. Must be non-negative and less than the size of the collection.**
Parameter name: index
   at System.ThrowHelper.ThrowArgumentOutOfRangeException(ExceptionArgument argument, ExceptionResource resource)
   at ScottPlot.Plottable.FinancePlot.GetAxisLimits()
   at ScottPlot.Settings.<>c.<AxisAutoX>b__53_4(IPlottable x)
   at System.Linq.Enumerable.WhereSelectEnumerableIterator`2.MoveNext()
   at System.Linq.Buffer`1..ctor(IEnumerable`1 source)
   at System.Linq.Enumerable.ToArray[TSource](IEnumerable`1 source)
   at ScottPlot.Settings.AxisAutoX(Int32 xAxisIndex, Double margin)
   at ScottPlot.Settings.AxisAutoUnsetAxes()
   at ScottPlot.Plot.Render(Bitmap bmp, Boolean lowQuality)
   at ScottPlot.Control.ControlBackEnd.Render(Boolean lowQuality, Boolean skipIfCurrentlyRendering)
   at ScottPlot.FormsPlot.Render(Boolean lowQuality, Boolean skipIfCurrentlyRendering)
   at DBVisor.DbVisorForm.RenderBinance(String database, String tableName) in d:\Code\DBVisor\DbVisorForm.cs:line 163
